Redshift Split Casting Issue

### Trino version

482

### Please describe the bug

I am getting the following pretty regularly on queries after enabling the unloading feature in Redshift connections.

```
java.util.concurrent.CompletionException: java.lang.ClassCastException: class io.trino.plugin.redshift.RedshiftUnloadSplit cannot be cast to class io.trino.plugin.jdbc.JdbcSplit (io.trino.plugin.redshift.RedshiftUnloadSplit and io.trino.plugin.jdbc.JdbcSplit are in unnamed module of loader io.trino.server.PluginClassLoader @3577d2a2)
	at java.base/java.util.concurrent.CompletableFuture.wrapInCompletionException(CompletableFuture.java:323)
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:359)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:364)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:670)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:531)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1831)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
	at java.base/java.lang.Thread.run(Thread.java:1474)
Caused by: java.lang.ClassCastException: class io.trino.plugin.redshift.RedshiftUnloadSplit cannot be cast to class io.trino.plugin.jdbc.JdbcSplit (io.trino.plugin.redshift.RedshiftUnloadSplit and io.trino.plugin.jdbc.JdbcSplit are in unnamed module of loader io.trino.server.PluginClassLoader @3577d2a2)
	at io.trino.plugin.jdbc.DynamicFilteringJdbcSplitSource.lambda$getNextBatch$1(DynamicFilteringJdbcSplitSource.java:104)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:214)
	at java.base/java.util.Collections$2.tryAdvance(Collections.java:5182)
	at java.base/java.util.Collections$2.forEachRemaining(Collections.java:5190)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:570)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:560)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:723)
	at io.trino.plugin.jdbc.DynamicFilteringJdbcSplitSource.lambda$getNextBatch$0(DynamicFilteringJdbcSplitSource.java:110)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:667)
	... 5 more
```

Queries that are triggering this look like:

```sql
WITH some_cte as (
  SELECT value
  FROM redshift.public.table
  WHERE reference_value = 'some_string'
)
SELECT value
FROM postgres.public.table
WHERE value IN (
  SELECT value
  FROM upstream_values
  )
```

So it looks like this is happening when using a CTE querying Redshift in a select statement reading from postgres. I rolled back to 481 and I do not see this issue.

--- a/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DynamicFilteringJdbcSplitSource.java
+++ b/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DynamicFilteringJdbcSplitSource.java
@@ -101,7 +101,10 @@ public class DynamicFilteringJdbcSplitSource
                 .thenApply(splits -> splits.stream()
                         // attach dynamic filter constraint to JdbcSplit
                         .map(split -> {
-                            JdbcSplit jdbcSplit = (JdbcSplit) split;
+                            // Delegates that want to consume the dynamic filter must produce JdbcSplit or a subclass.
+                            if (!(split instanceof JdbcSplit jdbcSplit)) {
+                                return split;
+                            }
                             // If split was a subclass of JdbcSplit, there would be additional information
                             // that we would need to pass further on.
                             verify(jdbcSplit.getClass() == JdbcSplit.class, "Unexpected split type %s", jdbcSplit);

--- a/plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcDynamicFilteringSplitManager.java
+++ b/plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcDynamicFilteringSplitManager.java
@@ -140,4 +140,33 @@ public class TestJdbcDynamicFilteringSplitManager
         assertThat(capturedTableHandle.get()).isNull();
         splitSource.close();
     }
+
+    @Test
+    public void testGetNextBatchPassesThroughNonJdbcSplits()
+            throws Exception
+    {
+        ConnectorSplit nonJdbcSplit = new ConnectorSplit() {};
+        JdbcDynamicFilteringSplitManager manager = new JdbcDynamicFilteringSplitManager(
+                new ConnectorSplitManager()
+                {
+                    @Override
+                    public ConnectorSplitSource getSplits(ConnectorTransactionHandle transaction, ConnectorSession session, ConnectorTableHandle table, Set<ColumnHandle> dynamicFilterColumns, Constraint constraint)
+                    {
+                        return new FixedSplitSource(ImmutableList.of(nonJdbcSplit));
+                    }
+                });
+        Set<ColumnHandle> dynamicFilterColumns = ImmutableSet.of(new TestColumnHandle());
+        ConnectorSplitSource splitSource = manager.getSplits(
+                TRANSACTION_HANDLE,
+                SESSION,
+                TABLE_HANDLE,
+                dynamicFilterColumns,
+                alwaysTrue());
+
+        TupleDomain<ColumnHandle> predicate = TupleDomain.withColumnDomains(
+                ImmutableMap.of(JDBC_COLUMN_HANDLE, Domain.singleValue(BIGINT, 42L)));
+        List<ConnectorSplit> batch = splitSource.getNextBatch(100, new DynamicFilterSnapshot(predicate, true)).get();
+        assertThat(batch).containsExactly(nonJdbcSplit);
+        splitSource.close();
+    }
 }

--- a/plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftUnload.java
+++ b/plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftUnload.java
@@ -18,6 +18,7 @@ import io.trino.Session;
 import io.trino.operator.OperatorInfo;
 import io.trino.testing.AbstractTestQueryFramework;
 import io.trino.testing.QueryRunner;
+import io.trino.testing.QueryRunner.MaterializedResultWithPlan;
 import io.trino.testing.sql.SqlExecutor;
 import io.trino.testing.sql.TestTable;
 import org.junit.jupiter.api.Test;
@@ -34,6 +35,7 @@ import java.util.Map;
 
 import static com.google.common.collect.ImmutableList.toImmutableList;
 import static com.google.common.collect.MoreCollectors.onlyElement;
+import static io.trino.SystemSessionProperties.ENABLE_DYNAMIC_FILTERING;
 import static io.trino.plugin.redshift.RedshiftQueryRunner.IAM_ROLE;
 import static io.trino.plugin.redshift.TestingRedshiftServer.TEST_SCHEMA;
 import static io.trino.testing.TestingNames.randomNameSuffix;
@@ -147,6 +149,36 @@ final class TestRedshiftUnload
                     assertThat(operatorInfo).isNull();
                 },
                 results -> assertThat(results.getRowCount()).isEqualTo(0));
+    }
+
+    @Test
+    void testUnloadWithDynamicFilter()
+    {
+        Session session = Session.builder(getSession())
+                .setCatalogSessionProperty("redshift", "join_pushdown_enabled", "false")
+                .build();
+
+        // Dynamic filter applied to the UNLOAD (probe) side of a join must not fail when attaching the predicate to the split
+        String query = "SELECT n.name FROM nation n JOIN nation r ON n.nationkey = r.regionkey WHERE r.name = 'ALGERIA'";
+        assertQuery(session, query, "VALUES ('ALGERIA')");
+
+        // The dynamic filter must actually reduce the rows Redshift unloads from the probe side
+        Session dynamicFilteringDisabled = Session.builder(session)
+                .setSystemProperty(ENABLE_DYNAMIC_FILTERING, "false")
+                .build();
+        assertThat(physicalInputPositions(session, query))
+                .isLessThan(physicalInputPositions(dynamicFilteringDisabled, query));
+    }
+
+    private long physicalInputPositions(Session session, String query)
+    {
+        QueryRunner queryRunner = getDistributedQueryRunner();
+        MaterializedResultWithPlan result = queryRunner.executeWithPlan(session, query);
+        return queryRunner.getCoordinator()
+                .getQueryManager()
+                .getFullQueryInfo(result.queryId())
+                .getQueryStats()
+                .getPhysicalInputPositions();
     }
 
     @Test

